snapshot.go

```diff
@@ -128,14 +129,36 @@ func (db *DB) snapshot(ctx context.Context, async bool, onSuccess func()) {
 		return
 	}
 
-	tx := db.beginRead()
+	tx, _, commit := db.begin()
```

Comment:
No, because we would then need a way to stop all compactions. Because otherwise what can happen is 

1. Snapshot begingRead = tx:5
2. Write tx:6 completes and triggers compaction.
3. Compaction happens and scoops up tx:6 and stores it in the next LSM level
4. Snapshot continues and snapshots the LSM causing the snapshot to contain a part that is > the snapshot tx

If we get a write tx then compactions wont be able to compact anything > the snapshot tx

Reply:
Maybe we should add to `pendingReadersWg` and have compaction wait for that? It seems incorrect to have snapshots grab a write txn (even though we've done that historically) just to control the concurrency with snapshots; feels like the wrong tool for the job.

Also, what happens if we do have normal reads during a compaction? Could we possibly miss data while it's being compacted?

Reply:
`pendingReadersWg` is meant only for the block level not the index level, so that's not the right tool for the job, I also would like to avoid blocking snapshots on reads if possible.

The index operations are atomic, so reads during compaction will either get the uncompacted data or the compacted data, there's no timing where the data isn't a part of the index.

Reply:
> pendingReadersWg is meant only for the block level not the index level, so that's not the right tool for the job

But why make a distinction? There's a 1:1 mapping. I'm also wondering more about this since running into a deadlock in https://github.com/polarsignals/frostdb/issues/741 and a little confused why we need concurrency control (i.e. waitgroups) on individual parts as well.

> I also would like to avoid blocking snapshots on reads if possible.

Did you mean compactions? We wouldn't be blocking snapshots since they would be just like any other read. I guess we would be blocking compactions, but wouldn't the original problem be solved if we had snapshot isolation so that reads don't read a part with a txn > than the snapshots txn in which case we don't need to block anyone?

Reply:
We need concurrency control on parts because we can't know if they're apart of a read and need to correctly determine when they can be released. So either arrow memory can reclaim it or the parquet file can be cleaned up.

Reply:
Coming back to this, I think this is something we need to resolve.

How come beginning a write txn for a snapshot blocks compactions? I might be missing something, but I don't see how it does.

I'm leaning more towards snapshots acquiring compaction locks. It definitely feels very bad either way. Could we do something wrt coordinating compaction and snapshot txn ids? Isn't the compaction problem something that would happen for any read that would violate snapshot isolation for reads if there's a concurrent compaction?

Reply:
> How come beginning a write txn for a snapshot blocks compactions? I might be missing something, but I don't see how it does.

Because compaction waits for the watermark to reach the highest tx in the level. Meaning if we have a write tx open for snapshotting; and a new write comes in and completes, compaction will sit and wait for the watermark to reach that level which wont happen until the snapshot completes.

I don't think having snapshots acquire the compaction lock is the correct thing. There's then also a weird interaction where a new table is created during a snapshot, but after the locks are acquired. It's just not the correct way to coordinate those things. That lock is just to prevent multiple compactions and isn't meant to leak outside the index layer.

Reply:
Let's discuss in the storage meeting. Maybe the write txn is the lesser of two evils, but it still doesn't solve the issue for vanilla read txns.

Reply:
What we could do is have there be a database level snapshot mutex. Where snapshots grab the write lock, and all compactions grab the read lock. This way no compactions are happening while a snapshot is taking place.

Reply:
Also what issue with vanilla read transactions?